[IMP] l10n_de: add a logic for the different accounts according to ta…

jco-odoo opened the pull request:
…x rates

Before, when only invoicing is installed (e.g. in case of the PoS), a
not so easy to surpass error would be triggered when using products
with a lower tax rate, e.g. drinks are only at 7% instead of 19%.
Because there is a constraint, upon validation of the invoice, that the
accounts used must correspond with the tax rate and you can not set a
specific account in case of only invoicing, an error will be raised when
you try to validate the generated invoice.

We solve it by inheriting the method searching for the product accounts and
when no income/expense account on the product is set, but a tax is, to suggest
the account corresponding to the tax and its rate.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

jco-odoo commented:
@smetl Any suggestions on this one?

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/62743).

oco-odoo commented on addons/l10n_de/models/datev.py:
```diff
@@ -33,3 +33,25 @@ def _post(self, soft=True):
+            if not self.property_account_income_id:
+                # Search for a corresponding account
+                income_account = self.env['account.account'].search([('internal_group', '=', 'income'), ('deprecated', '=', False),
+                                                               ('tax_ids', 'in', self.taxes_id.filtered(lambda t: t.company_id == self.env.company).ids)], limit=1)
```
You should call filtered only once before entering the condition and store it into a variable.

So, I'd reshape the whole condition like this

```suggestion
if self.env.company.country_id.code == "DE" and not self.property_account_income_id:
    company_taxes = self.taxes_id.filtered(lambda t: t.company_id == self.env.company)
    if company_taxes:
        ...
```

oco-odoo commented on addons/l10n_de/models/datev.py:
```diff
@@ -33,3 +33,25 @@ def _post(self, soft=True):
+            if not self.property_account_expense_id:
+                # Search for a corresponding account
+                expense_account = self.env['account.account'].search([('internal_group', '=', 'expense'), ('deprecated', '=', False),
+                                                               ('tax_ids', 'in', self.supplier_taxes_id.filtered(lambda t: t.company_id == self.env.company).ids)], limit=1)
```
Same here

oco-odoo reviewed (changes requested):
I think there is an easier and more user-friendly way to do it. Here, it's kind of magical and if people don't know it's there, they have no way to expect it'll work like that. 

Can't we simply put an onchange on the tax fields and set property_account_income_id and property_account_expense_id from it if those fields have no value yet? The only drawback with that is that the accounts won't be set if we assign the taxes to the product before setting them on the accounts, but is it really going to happen?

jco-odoo commented:
@oco-odoo I think that onchanging the stuff would only bring multi-company problems (property) and it would not solve the problem on existing dbs.  It only triggers when you did not explicitly set the account on the product.  The default taxes on the accounts are already set when installing the CoA.  It is also a bit magical that on confirmation of an invoice, certainly if you only have invoicing, the thing starts to complain about accounts and taxes not being matched.  Another option I see, is to do the super and only when the csn raise would be triggered, that we change the account.  (that way, we avoid any unwanted behaviour)  

The default taxes on accounts, seems to be related to Germany mostly.  But I suppose that most people are concentrated on putting the right tax.

jco-odoo commented:
@oco-odoo I changed the logic to make sure it is only applied in order to avoid violating the constraint if it was configured badly. Maybe we should also only apply when property_... on the product is not set.  (so it was not configured at all)

I also added a commit with a small fix, so if you would add an invoice line from another company, it would still put the correct taxes.

oco-odoo commented on addons/account/models/account_move.py:
```diff
@@ -3113,7 +3113,7 @@ def _get_computed_taxes(self):
 
         if self.company_id and tax_ids:
-            tax_ids = tax_ids.filtered(lambda tax: tax.company_id == self.company_id)
+            tax_ids = tax_ids.filtered(lambda tax: tax.company_id == self.move_id.company_id)
```
Isn't self.company_id related to the move ? Also, is it the right version to do it? It feels like something that should be done in 13.0.

oco-odoo commented on addons/l10n_de/models/datev.py:
```diff
@@ -33,3 +33,23 @@ def _post(self, soft=True):
+                result['income'] = self.env['account.account'].search([('internal_group', '=', 'income'), ('deprecated', '=', False),
+                                                               ('tax_ids', 'in', taxes.ids)], limit=1)
+            supplier_taxes = self.supplier_taxes_id.filtered(lambda t: t.company_id == company)
+            if result.get('expense') and result['expense'].tax_ids and supplier_taxes and supplier_taxes[0] not in result['expense'].tax_ids:
```
If you have no income account but have an expense account, you will never reach that condition, as company will be False.

oco-odoo reviewed (changes requested):
I find it weird to change the logic so that we will override income and expense account only when they are set from the product or product category. Shouldn't it rather be the opposite ? If they are not explicitly configured to be set to certain values, we force them so that they will match the constraint. 

So it means the constraint will break if the configuration is bad ... but I think that's what a bad configuration should do, instead of trying to make it work silently. Also, it's more robust: if in the future, the constraint gets removed before of some corner case that we need to support, we'll just remove it without changing any behavior elsewhere, and hence disturbing the users way less.

jco-odoo commented:
@oco-odoo I hope it is ok now

oco-odoo commented:
@robodoo r+
